airsonic-main/src/main/java/org/airsonic/player/dao/LegacyHsqlDaoHelper.java

```diff
@@ -31,29 +51,226 @@ public void checkpoint() {
+            LOG.info("Database connection established. Current version is: {}.", conn.getMetaData().getDatabaseProductVersion());
+            // On upgrade, the official documentation recommends that we
+            // run 'SHUTDOWN SCRIPT' to compact all the database into a
+            // single SQL file.
+            //
+            // In practice, if we don't do that, we did not observe issues
+            // immediately but after the upgrade.
+            LOG.info("Shutting down database...");
+            try (Statement st = conn.createStatement()) {
+                st.execute("SHUTDOWN SCRIPT");
+            }
+        }
+
+        LOG.info("Database connection closed. Current version is: {}.", getHsqldbDatabaseVersion());
+    }
+
+    /**
+     * Checks if the database is correctly shut down
+     */
+    private boolean isHsqldbDatabaseCorrectlyShutdown() {
+
+        // The db/airsonic.data file may or may not exist (contains CACHED tables).
+        // The db/airsonic.backup file may or may not exist (not used during normal operation).
+
+        // The db/airsonic.properties file must exist (contains db properties).
+        if (!new File(settingsService.getDefaultJDBCPath() + ".properties").exists()) {
+            LOG.warn("The db/airsonic.properties file does not exist. The database may be corrupt.");
+            return false;
+        }
+        // The db/airsonic.script file must exist (contains db data from all tables except CACHED).
+        if (!new File(settingsService.getDefaultJDBCPath() + ".script").exists()) {
+            LOG.warn("The db/airsonic.script file does not exist. The database may be corrupt.");
+            return false;
+        }
+        // The db/airsonic.log file should not exist (contains db logs that are not supported in the upgrade).
+        if (new File(settingsService.getDefaultJDBCPath() + ".log").exists()) {
+            LOG.warn("The db/airsonic.log file exists. The database may not be shut down correctly.");
+            return false;
+        }
+        // The db/airsonic.lck file should not exist (lock file, indicates another running process or unclean shutdown).
+        if (new File(settingsService.getDefaultJDBCPath() + ".lck").exists()) {
+            LOG.warn("The db/airsonic.lck file exists. The database may not be shut down correctly.");
+            return false;
+        }
+
+        return true;
+    }
+
+    private void upgradeHsqldbDatabaseSafely() {
+        try {
+            // Check that it is safe to do the upgrade
+            if (!isHsqldbDatabaseCorrectlyShutdown()) {
+                LOG.warn(
+                    "Database has not been shut down properly. " +
+                    "Airsonic should be stopped properly before attempting to upgrade the database.");
+            }
+
+            // Perform the upgrade
+            Path backupPath = performHsqldbDatabaseBackup();
```

Comment:
I think this might fail if the database's lock file is being held

Reply:
You're right. A big assumption of `performHsqldbDatabaseBackup` is that you are not running it in parallel with any other connection.

Reply:
Fixed by moving this to `Application`.